[css-grid] Accumulate nested subgrids' margin, border, and padding to base size of outer grid tracks.

sammygill opened the pull request:
#### ebd030f9b8667698930ad2ffa5de7c56bbbab8cd
<pre>
[css-grid] Accumulate nested subgrids&apos; margin, border, and padding to base size of outer grid tracks.
<a href="https://bugs.webkit.org/show_bug.cgi?id=260531">https://bugs.webkit.org/show_bug.cgi?id=260531</a>
rdar://114271839

Reviewed by Matt Woodrow.

Currently, if there is a subgrid we will take the sum of its margin,
border, and padding (mbp) at each edge and apply it to the track sizing
algorithm for the track the edge is in. This sum may then get applied to
the base size of the track if needed.

However, if there are nested subgrids, we need to take them into account
and accumulate their values as well so that any adjacent chunks of mbp
are considered as a single layer when sizing the track.

We can accomplish this by keeping track of a variable that stores the
accumulated value as we recurse through nested subgrids. This value will
first be set to 0 in the first call to accumulateIntrinsicSizesForTrack.
As we iterate over the track&apos;s items, if we find a subgrid we will
compute the margin, border, and padding values it contributes to the
track and add it to the variable we are using to store the accumulated
value and potentially update the track size.

The key idea is that when we recurse into a new level of a nested
subgrid we will pass along and use this updated accumulated value with
the subgrid&apos;s mbp, but as we traverse through subgrids within the same
level will: use the same accumulated value, add the subgrid&apos;s mbp to
the value, and then potentially update the track size.

* LayoutTests/TestExpectations:
* Source/WebCore/rendering/GridTrackSizingAlgorithm.cpp:
(WebCore::computeSubgridMarginBorderPadding):
(WebCore::GridTrackSizingAlgorithm::accumulateIntrinsicSizesForTrack):
(WebCore::GridTrackSizingAlgorithm::resolveIntrinsicTrackSizes):
(WebCore::addSubgridMarginBorderPadding): Deleted.
* Source/WebCore/rendering/GridTrackSizingAlgorithm.h:

Canonical link: <a href="https://commits.webkit.org/268947@main">https://commits.webkit.org/268947@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/bba87833c88678fb1ef19cd04e204675bf61543d

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/21099 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/26/builds/21474 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/22165 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/22979 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/19628 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/21339 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/23/builds/24732 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/16/builds/21664 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/34/builds/20871 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/21323 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/15/builds/21044 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/18294 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/23834 "Built successfully") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/39/builds/18202 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/19122 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/25406 "Passed tests") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/19283 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/19324 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/23339 "Passed tests") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/7/builds/19874 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/17/builds/16891 "Passed tests") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/30/builds/19151 "Built successfully") | | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/5064 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/4/builds/23444 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/31/builds/19726 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

mattwoodrow commented on Source/WebCore/rendering/GridTrackSizingAlgorithm.cpp:
```diff
@@ -1430,13 +1425,28 @@ void GridTrackSizingAlgorithm::resolveIntrinsicTrackSizes()
+                const auto oppositeAxis = m_direction == GridTrackSizingDirection::ForColumns ? GridTrackSizingDirection::ForRows : GridTrackSizingDirection::ForColumns;
+                Vector<LayoutUnit> accumulatedSubgridMarginBorderPaddingsInTrack(m_renderGrid->numTracks(oppositeAxis));
+
+                for (const auto& subgrid : subgridsWithEdgesInTrack) {
```
I'm not quite sure I follow the logic here.

`subgridsWithEdgesInTrack` is the set all of sub grids that have a start/end track that is also this track. Two entries in that set might be a parent-child relationship (for a nested sub grid), or siblings.

We're taking the sum of the paddings for all sub grids that start in the same track in the opposite axis, and then take the max of those.

I don't understand why the opposite axis (and the sub grids start line in it) is the boundary at which we sum vs max.

mattwoodrow replied:
Is the assumption that if two entries share the same start line in the opposite axis, then they must be a parent-child/nested-subgrid relationship?

I don't think that's guaranteed, you can set forced grid placement on items to make multiple items share the same grid cell.

sammygill replied:
Yeah I was attempting to use the opposite axis index as a way of determining how subgrids were nested, but that was not quite correct as your example shows. 


I went back and slightly modified the approach. Instead of creating a list of subgrids we can just maintain a variable that accumulates as we nest into subgrids. It seems like the current recursive nature of the calls to `accumulateIntrinsicSizesForTrack` correctly captures the nested subgrid structure.

Let me know what you think!

sammygill committed:
[css-grid] Accumulate nested subgrids' margin, border, and padding to base size of outer grid tracks.
https://bugs.webkit.org/show_bug.cgi?id=260531
rdar://114271839

Reviewed by Matt Woodrow.

Currently, if there is a subgrid we will take the sum of its margin,
border, and padding (mbp) at each edge and apply it to the track sizing
algorithm for the track the edge is in. This sum may then get applied to
the base size of the track if needed.

However, if there are nested subgrids, we need to take them into account
and accumulate their values as well so that any adjacent chunks of mbp
are considered as a single layer when sizing the track.

We can accomplish this by keeping track of a variable that stores the
accumulated value as we recurse through nested subgrids. This value will
first be set to 0 in the first call to accumulateIntrinsicSizesForTrack.
As we iterate over the track's items, if we find a subgrid we will
compute the margin, border, and padding values it contributes to the
track and add it to the variable we are using to store the accumulated
value and potentially update the track size.

The key idea is that when we recurse into a new level of a nested
subgrid we will pass along and use this updated accumulated value with
the subgrid's mbp, but as we traverse through subgrids within the same
level will: use the same accumulated value, add the subgrid's mbp to
the value, and then potentially update the track size.

* LayoutTests/TestExpectations:
* Source/WebCore/rendering/GridTrackSizingAlgorithm.cpp:
(WebCore::computeSubgridMarginBorderPadding):
(WebCore::GridTrackSizingAlgorithm::accumulateIntrinsicSizesForTrack):
(WebCore::GridTrackSizingAlgorithm::resolveIntrinsicTrackSizes):
(WebCore::addSubgridMarginBorderPadding): Deleted.
* Source/WebCore/rendering/GridTrackSizingAlgorithm.h:

Canonical link: https://commits.webkit.org/268947@main
